src/java.base/share/classes/sun/net/www/protocol/jrt/JavaRuntimeURLConnection.java

```diff
@@ -28,117 +28,92 @@
+     * connection as "connected".
      */
-    private static Resource findResource(String module, String name) {
-        if (reader != null) {
-            URL url = toJrtURL(module, name);
-            ImageLocation location = reader.findLocation(module, name);
-            if (location != null) {
-                return new Resource() {
-                    @Override
-                    public String getName() {
-                        return name;
-                    }
-                    @Override
-                    public URL getURL() {
-                        return url;
-                    }
-                    @Override
-                    public URL getCodeSourceURL() {
-                        return toJrtURL(module);
-                    }
-                    @Override
-                    public InputStream getInputStream() throws IOException {
-                        byte[] resource = reader.getResource(location);
-                        return new ByteArrayInputStream(resource);
-                    }
-                    @Override
-                    public int getContentLength() {
-                        long size = location.getUncompressedSize();
-                        return (size > Integer.MAX_VALUE) ? -1 : (int) size;
-                    }
-                };
+    private synchronized Node connectResourceNode() throws IOException {
+        if (resourceNode == null) {
+            if (module.isEmpty() || path == null) {
+                throw new IOException("cannot connect to jrt:/" + module);
             }
+            Node node = READER.findNode("/modules/" + module + "/" + path);
+            if (node == null || !node.isResource()) {
+                throw new IOException(module + "/" + path + " not found");
+            }
+            this.resourceNode = node;
+            super.connected = true;
         }
-        return null;
+        return resourceNode;
     }
 
     @Override
-    public synchronized void connect() throws IOException {
-        if (!connected) {
-            if (name == null) {
-                String s = (module == null) ? "" : module;
-                throw new IOException("cannot connect to jrt:/" + s);
-            }
-            resource = findResource(module, name);
-            if (resource == null)
-                throw new IOException(module + "/" + name + " not found");
-            connected = true;
-        }
+    public void connect() throws IOException {
+        connectResourceNode();
     }
 
     @Override
     public InputStream getInputStream() throws IOException {
-        connect();
-        return resource.getInputStream();
+        return new ByteArrayInputStream(READER.getResource(connectResourceNode()));
     }
 
     @Override
     public long getContentLengthLong() {
+        // Note: UncheckedIOException is thrown by the Node subclass in
+        // ExplodedImage (this not obvious, so worth calling out).
```

Comment:
This protocol handler is for images build, not exploded builds. ImageReaderFactory.getImageReader will fail if you attempt it with an exploded build. So I'm not sure yet why this comment is here.

Reply:
@RogerRiggs The jrt protocol handler is for images builds. Any attempt, in exploded build, to connect to a jrt resource will blow up as ImageReaderFactory.getImageReader fails.

Reply:
Right, it would never occur. Remove as unnecessary.